### DISCUSSION TITLE
Cannot brew install ruby on CentOS

### DISCUSSION THREAD

**Discussion by PiegoIlTempo** [Homebrew/discussions#3421] [Linux] [answered]
Hello, 

for some reason I'm not able to build ruby on linux CentOS, and I get the following error:

```
==> ./configure --prefix=/var/www/.linuxbrew/Cellar/ruby/3.1.2 --enable-shared --with-sitedir=/var/www/.linuxbrew/lib/ruby/site_ruby --with-vendordir=/var/ww
==> make
==> make install
Last 15 lines from /var/www/.cache/Homebrew/Logs/ruby/03.make:
from /tmp/ruby-20220710-5707-1x47ene/ruby-3.1.2/lib/rubygems/ext/builder.rb:195:in `block in build_extensions'
from /tmp/ruby-20220710-5707-1x47ene/ruby-3.1.2/lib/rubygems/ext/builder.rb:192:in `each'
from /tmp/ruby-20220710-5707-1x47ene/ruby-3.1.2/lib/rubygems/ext/builder.rb:192:in `build_extensions'
from /tmp/ruby-20220710-5707-1x47ene/ruby-3.1.2/lib/rubygems/installer.rb:853:in `build_extensions'
from /tmp/ruby-20220710-5707-1x47ene/ruby-3.1.2/lib/rubygems/installer.rb:328:in `install'
from ./tool/rbinstall.rb:899:in `block in install'
from ./tool/rbinstall.rb:713:in `no_write'
from ./tool/rbinstall.rb:899:in `install'
from ./tool/rbinstall.rb:1068:in `block (2 levels) in <main>'
from ./tool/rbinstall.rb:1043:in `foreach'
from ./tool/rbinstall.rb:1043:in `block in <main>'
from ./tool/rbinstall.rb:1127:in `block in <main>'
from ./tool/rbinstall.rb:1124:in `each'
from ./tool/rbinstall.rb:1124:in `<main>'
make: *** [do-install-all] Error 1

READ THIS: https://docs.brew.sh/Troubleshooting
```

Also thank you @danielnachun for the support and the suggestion to open this discussion here.

As he said [here](https://github.com/Homebrew/brew/issues/13535#issuecomment-1179776675) :
`It appears to be because the native extensions don't use brewed glibc when they should be. `

But I have no idea on what to try in order to solve this.

**Comment by SMillerDev** [comment]
Can you post the full output of `/var/www/.cache/Homebrew/Logs/ruby/03.make`?

**Reply by PiegoIlTempo** [reply]
> Can you post the full output of `/var/www/.cache/Homebrew/Logs/ruby/03.make`?

Since it is very long, I'll attach the file itself (I just added .txt to be able to upload it).

[03.make.txt](https://github.com/Homebrew/discussions/files/9079892/03.make.txt)

**Reply by SMillerDev** [reply]
> current directory: /var/www/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension
/usr/bin/make DESTDIR\\=
/usr/bin/make: relocation error: /var/www/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference

> make failed, exit code 127

This seems to be the actual error. @Homebrew/Linux any ideas?

**Comment by danielnachun** [comment]
Just an update from me so far - I have some suspicion that a shim bypass is occurring here.  Our `gcc`/`ld` shim should normally make sure that the linker finds brewed `glibc` first, but the relevant error I get (which is identical to what @PiegoIlTempo posted) seems to indicate that something weird is happening with `glibc`:
```
Building native extensions. This could take a while...
/tmp/ruby-20220710-15949-gq0040/ruby-3.1.2/lib/rubygems/ext/builder.rb:95:in `run': ERROR: Failed to build gem native extension. (Gem::Ext::BuildError)

    current directory: /home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension
/home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/bin/ruby --disable\\=gems -I/home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/3.1.0/x86_64-linux -I /tmp/ruby-20220710-15949-gq0040/ruby-3.1.2/lib -r ./siteconf20220710-41954-szn54v.rb extconf.rb
creating Makefile

current directory: /home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension
/home/dnachun/.linuxbrew/opt/make/bin/make DESTDIR\\= clean
make[1]: Entering directory '/home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension'
rm: relocation error: /home/dnachun/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference
make[1]: [Makefile:193: clean-static] Error 127 (ignored)
/bin/sh: relocation error: /home/dnachun/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference
make[1]: [Makefile:176: clean] Error 127 (ignored)
make[1]: Leaving directory '/home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension'

current directory: /home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension
/home/dnachun/.linuxbrew/opt/make/bin/make DESTDIR\\=
make[1]: Entering directory '/home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension'
echo: relocation error: /home/dnachun/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference
make[1]: *** [Makefile:245: constants.o] Error 127
make[1]: *** Waiting for unfinished jobs....
echo: relocation error: /home/dnachun/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference
make[1]: *** [Makefile:245: lexer.o] Error 127
echo: relocation error: /home/dnachun/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference
make[1]: *** [Makefile:245: lexstate.o] Error 127
echo: relocation error: /home/dnachun/.linuxbrew/lib/libpthread.so.0: symbol __libc_vfork, version GLIBC_PRIVATE not defined in file libc.so.6 with link time reference
make[1]: *** [Makefile:245: location.o] Error 127
make[1]: Leaving directory '/home/dnachun/.linuxbrew/Cellar/ruby/3.1.2/lib/ruby/gems/3.1.0/gems/rbs-2.1.0/ext/rbs_extension'

make failed, exit code 2
```
Another possibility is that there is a pre-compiled binary sneaking in here somewhere that is somehow pulling in linkage to the host `glibc`.  I don't know the Ruby-native code interface as well as I do for Python or Perl so it may take some digging around to figure what's really going on here.  I do think this is worth fixing because I'm fairly sure that at the moment Ruby is one of the few formulae that does **not** work properly with `HOMEBREW_RELOCATE_BUILD_PREFIX=1`.

**Reply by danielnachun** [reply]
Now this is odd - I don't know how I missed this before, but the failures actually seem to be with `rm`, `/bin/sh`, `echo` etc.  Those aren't even provided by `brew` unless the right packages are installed (i.e. `coreutils`) so that makes this whole thing even more weird.

**Reply by PiegoIlTempo** [reply]
And could installing them through brew fix this? Since those would be under brew's control now.

I would like being more helpful but I have no experience with these kind of things especially with Ruby since I'm installing it only because it is a dependency to another project. Let me know if there is anything I can do (or try to do) to help.

**Reply by CyCle1024** [reply]
@danielnachun Hi, I met this problem exactly PiegoIlTempo and you posted. I have tried `HOMEBREW_RELOCATE_BUILD_PREFIX=1.` and it won't fix. Any update on this ruby glibc linking problem by now?
btw, the command `rm`, `echo` etc. works properly in my environment.

**Reply by PiegoIlTempo** [reply]
> @danielnachun Hi, I met this problem exactly PiegoIlTempo and you posted. I have tried `HOMEBREW_RELOCATE_BUILD_PREFIX=1.` and it won't fix. Any update on this ruby glibc linking problem by now? btw, the command `rm`, `echo` etc. works properly in my environment.

Hello, actually from my understanding Brew really needs to binary of gcc in the root bin folder, if it's not there you could try downloading an extrernal gcc binary and use that one to install gcc@5 with Brew as Daniel suggested [here ](https://github.com/Homebrew/brew/issues/13535#issuecomment-1179672786) but I didn't test it.

I was able to solve this in my environment because it allowed to install very specific packages as sudo, and it just happened that one of those was gcc, [here ](https://stackoverflow.com/questions/72951813/how-to-install-homebrew-on-jelastic) is my solution anyway.

Hope this helps.

**Reply by CyCle1024** [reply]
> > @danielnachun Hi, I met this problem exactly PiegoIlTempo and you posted. I have tried `HOMEBREW_RELOCATE_BUILD_PREFIX=1.` and it won't fix. Any update on this ruby glibc linking problem by now? btw, the command `rm`, `echo` etc. works properly in my environment.
> 
> Hello, actually from my understanding Brew really needs to binary of gcc in the root bin folder, if it's not there you could try downloading an extrernal gcc binary and use that one to install gcc@5 with Brew as Daniel suggested [here ](https://github.com/Homebrew/brew/issues/13535#issuecomment-1179672786) but I didn't test it.
> 
> I was able to solve this in my environment because it allowed to install very specific packages as sudo, and it just happened that one of those was gcc, [here ](https://stackoverflow.com/questions/72951813/how-to-install-homebrew-on-jelastic) is my solution anyway.
> 
> Hope this helps.

Thanks for your reply. Actually I have successfully installed gcc by `brew install gcc` along with `glibc` built from source, everything works fine right now except for ruby. But what you suggest gives me a idea: “I should not use brew glibc and gcc to build ruby”, I unlink brew gcc and glibc, and I use host gcc and glibc building brew ruby from source correctly! Then I link brew gcc and glibc again. This is a trick, but I hope that the brew glibc could have good compatibility with host glibc.

**Reply by PiegoIlTempo** [reply]
I think I had faced a very similar problem installing Ruby, and I solved it by doing brew install gcc@5, idk why but building Ruby with the latest brew gcc doesn't work, but if you have installed gcc@5 it does, maybe this was the probem, anyway glad you sorted it out!

**Reply by CyCle1024** [reply]
thx, maybe someone of the homebrew team can have a deep look into this problem if this is a serious problem for several source build formula.

**Reply by danielnachun** [reply]
This is definitely something I will be looking at although my first few attempts haven’t been successful.  My guess at the moment is that we need to fix `rbconfig.rb` so that it doesn’t bypass our compiler shim.  If my theory is right, the cause of the failure when building from source is that Ruby is being built with brewed GCC and glibc but the native gems try to use the host GCC and glibc and these are incompatible.  If I have a possible solution I’ll open a pull request with the fix and link it here!

**Reply by CyCle1024** [reply]
> This is definitely something I will be looking at although my first few attempts haven’t been successful. My guess at the moment is that we need to fix `rbconfig.rb` so that it doesn’t bypass our compiler shim. If my theory is right, the cause of the failure when building from source is that Ruby is being built with brewed GCC and glibc but the native gems try to use the host GCC and glibc and these are incompatible. If I have a possible solution I’ll open a pull request with the fix and link it here!

Thanks a lot !!!!

**Comment by PiegoIlTempo** [comment, marked as answer]
Ok, by my surprise this morning after executing a `brew update` and trying a `brew install ruby` I got the following message:
![image](https://user-images.githubusercontent.com/40637700/178428085-d3489f21-797e-4939-8c4b-46d59c696779.png)

Then I tried installing again the repo I need and now I received another error, but regarding fortran:

```
==> ./configure --prefix=/var/www/.linuxbrew/Cellar/open-mpi/4.1.4 --enable-ipv6 --enable-mca-no-build=reachable-netlink --with-libevent=/var/www/.linuxbrew/opt/libevent --with-sge
Last 15 lines from /var/www/.cache/Homebrew/Logs/open-mpi/01.configure:
checking for gfortran... gfortran
checking whether we are using the GNU Fortran compiler... yes
checking whether gfortran accepts -g... yes
checking whether ln -s works... yes
checking if Fortran compiler works... no
**********************************************************************
* It appears that your Fortran compiler is unable to produce working
* executables.  A simple test application failed to properly
* execute.  Note that this is likely not a problem with Open MPI,
* but a problem with the local compiler installation.  More
* information (including exactly what command was given to the
* compiler and what error resulted when the command was executed) is
* available in the config.log file in the Open MPI build directory.
**********************************************************************
configure: error: Could not run a simple Fortran program.  Aborting.

READ THIS: https://docs.brew.sh/Troubleshooting
```

I'm trying to troubleshoot this by myself, but I wanted to give you a quick update that this seems to be solved somehow, if you have any suggestion on how to fix this new problem as well it is welcomed.

Again, thank you to all of you for your help!

**Reply by danielnachun** [reply]
You might be able to get the rest of your dependencies installed by adding `export HOMEBREW_RELOCATE_BUILD_PREFIX=1`.  Almost all packages can be used from bottles this way, with some exceptions at the moment.  I'd try this first and then rebuild from source anything that seems broken.

**Reply by PiegoIlTempo** [reply]
I just noticed that this problem could be caused by [this](https://github.com/Homebrew/homebrew-core/issues/101919#issuecomment-1162740031) `brew install --force-bottle gcc` that I used, because gfortran was installed with gcc and even if I try to reinstall it I run in the same error I ran into when trying to install gcc, unless I force bottle here as well accomplishing nothing (since that is the cause of the problem I guess).

For now I will try to use `export HOMEBREW_RELOCATE_BUILD_PREFIX=1` to install the repo I need and let you know if it worked!

**Reply by PiegoIlTempo** [reply]
It worked out, thank you!

**Comment by Lightup1** [comment]
In my machine, after unlink gcc and glibc, brew install ruby successfully. 
The reanson is that the formulea somehow use system gcc.
Same error for openssl@3  that the formulea uses system perl and report lack of module and for wget brew reports make error just as ruby and can be fix by unlink gcc and glibc.
These seem to belong to the same error type that the formulea somehow use system build tools.

**Reply by danielnachun** [reply]
That's what I've been suspecting is the problem.  The solution I'm looking for is how to stop them form using the system GCC.  They shouldn't be bypassing our compiler shims and once we can figure out to stop this everything should work as expected.

**Comment by wscott** [comment]
Same problem here on SLES12
- Didn't have any luck unlinking gcc, gcc@5, & glib first because the system cc is too old to work
- can't use HOMEBREW_RELOCATE_BUILD_PREFIX because my prefix is too long

The ruby build is fine, but the native gem builds use the system CC which don't like libruby created with homebrew's gcc.

**Comment by wang-jinbo** [comment]
Got the same kind of problem, and didn't know how to fix
